File: server-java/src/main/java/com/jixiao2/server/home/HomeService.java
package com.jixiao2.server.home;

import com.jixiao2.server.employee.EmployeeService;
import com.jixiao2.server.menu.MenuPermissionService;
import java.sql.Timestamp;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

@Service
public class HomeService {

  private final JdbcTemplate jdbc;
  private final MenuPermissionService menuPermissionService;
  private final EmployeeService employeeService;

  public HomeService(
      JdbcTemplate jdbc, MenuPermissionService menuPermissionService, EmployeeService employeeService) {
    this.jdbc = jdbc;
    this.menuPermissionService = menuPermissionService;
    this.employeeService = employeeService;
  }

  /** 与绩效终审/校准能力一致：可看到「待终审」类待办的用户（非全员绩效列表）。 */
  private boolean canSeeFinalReviewTodos(String userId, String role) {
    if (userId == null || userId.isEmpty()) {
      return false;
    }
    if ("super_admin".equals(role)) {
      return true;
    }
    if (menuPermissionService.isMenuAllowed(userId, "performance_review_admin")) {
      return true;
    }
    return employeeService.readCalibrationAssigneeEmployeeIds().contains(userId);
  }

  private static int[] resolveYearMonth(Integer year, Integer month) {
    LocalDate now = LocalDate.now();
    int y = year != null ? year : now.getYear();
    int m = month != null ? Math.min(12, Math.max(1, month)) : now.getMonthValue();
    return new int[] {y, m};
  }

  private static Timestamp monthStart(int year, int month) {
    return Timestamp.valueOf(YearMonth.of(year, month).atDay(1).atStartOfDay());
  }

  private static Timestamp monthEndExclusive(int year, int month) {
    return Timestamp.valueOf(YearMonth.of(year, month).plusMonths(1).atDay(1).atStartOfDay());
  }

  public Map<String, Object> getTodos(String userId, Integer year, Integer month) {
    int[] ym = resolveYearMonth(year, month);
    Timestamp start = monthStart(ym[0], ym[1]);
    Timestamp endEx = monthEndExclusive(ym[0], ym[1]);
    String role = menuPermissionService.getUserRole(userId);

    List<Object> args = new ArrayList<Object>();
    List<String> whereParts = new ArrayList<String>();
    whereParts.add(
        "((pr.status IN ('template_selection','goal_setting') AND pr.employee_id=?) OR (pr.status=? AND pr.employee_id=?) OR "
            + "(pr.status=? AND (pr.manager_id=? OR pr.dotted_manager_id=?)) OR "
            + "(pr.status=? AND pr.employee_id=?) OR (pr.status=? AND pr.manager_id=?) OR "
            + "(pr.status=? AND (pr.manager_id=? OR pr.dotted_manager_id=?)) OR "
            + "(pr.status=? AND pr.dotted_manager_id=?) OR "
            + "(pr.status=? AND pr.employee_id=?))");
    args.add(userId);
    args.add("goal_rejected");
    args.add(userId);
    args.add("goal_pending_review");
    args.add(userId);
    args.add(userId);
    args.add("self_review");
    args.add(userId);
    args.add("manager_review");
    args.add(userId);
    args.add("dual_manager_review");
    args.add(userId);
    args.add(userId);
    args.add("dotted_manager_review");
    args.add(userId);
    args.add("issued");
    args.add(userId);
    if (canSeeFinalReviewTodos(userId, role)) {
      whereParts.add("(pr.status='final_review')");
    }

    StringBuilder sqlBuilder = new StringBuilder();
    sqlBuilder.append(
        "SELECT pr.id, pr.period, pr.status, pr.employee_id, eh.name AS employee_name, eh.department_name FROM performance_record pr "
            + "LEFT JOIN employee_hierarchy eh ON eh.employee_id = pr.employee_id WHERE (");
    sqlBuilder.append(String.join(" OR ", whereParts));
    sqlBuilder.append(
        ") AND pr.deleted_at IS NULL AND pr._updated_at >= ? AND pr._updated_at < ? ORDER BY pr._updated_at DESC");
    String sql = sqlBuilder.toString();
    args.add(start);
    args.add(endEx);

    List<Map<String, Object>> rows =
        jdbc.query(
            sql,
            args.toArray(),
            (rs, rn) -> {
              Map<String, Object> m = new LinkedHashMap<String, Object>();
              m.put("id", rs.getString("id"));
              m.put("period", rs.getString("period"));
              m.put("status", rs.getString("status"));
              m.put("employeeId", rs.getString("employee_id"));
              m.put("employeeName", rs.getString("employee_name"));
              m.put("departmentName", rs.getString("department_name"));
              return m;
            });

    List<Map<String, Object>> items = new ArrayList<Map<String, Object>>();
    for (Map<String, Object> r : rows) {
      String period = (String) r.get("period");
      String st = (String) r.get("status");
      String employeeName = (String) r.get("employeeName");
      String departmentName = (String) r.get("departmentName");
      String type = mapTodoType(st);
      String title = buildTitle(employeeName, st);
      Map<String, Object> it = new LinkedHashMap<String, Object>();
      it.put("id", r.get("id"));
      it.put("period", period);
      it.put("type", type);
      it.put("title", title);
      it.put("employeeId", r.get("employeeId"));
      it.put(
          "employeeName",
          employeeName != null && !employeeName.trim().isEmpty() ? employeeName.trim() : "");
      it.put(
          "departmentName",
          departmentName != null && !departmentName.trim().isEmpty() ? departmentName.trim() : "");
      items.add(it);
    }
    Map<String, Object> out = new LinkedHashMap<String, Object>();
    out.put("items", items);
    return out;
  }

  private static String mapTodoType(String status) {
    if ("template_selection".equals(status) || "goal_setting".equals(status)) {
      return "goal_setting";
    }
    if ("goal_rejected".equals(status)) {
      return "goal_rejected";
    }
    if ("goal_pending_review".equals(status)) {
      return "goal_pending_review";
    }
    if ("self_review".equals(status)) {
      return "self_review";
    }
    if ("manager_review".equals(status)) {
      return "manager_review";
    }
    if ("dual_manager_review".equals(status)) {
      return "dual_manager_review";
    }
    if ("dotted_manager_review".equals(status)) {
      return "dotted_manager_review";
    }
    if ("final_review".equals(status)) {
      return "final_review";
    }
    if ("issued".equals(status)) {
      return "issued";
    }
    return "self_review";
  }

  private static String buildTitle(String employeeName, String status) {
    String who =
        employeeName != null && !employeeName.trim().isEmpty() ? employeeName.trim() + " · " : "";
    if ("template_selection".equals(status) || "goal_setting".equals(status)) {
      return who + "目标设定";
    }
    if ("goal_rejected".equals(status)) {
      return who + "目标被驳回，请修改";
    }
    if ("goal_pending_review".equals(status)) {
      return who + "待审核目标";
    }
    if ("self_review".equals(status)) {
      return who + "绩效自评";
    }
    if ("manager_review".equals(status)) {
      return who + "直属上级评分";
    }
    if ("dual_manager_review".equals(status)) {
      return who + "上级与虚线上级评分";
    }
    if ("dotted_manager_review".equals(status)) {
      return who + "虚线上级评分";
    }
    if ("final_review".equals(status)) {
      return who + "待终审/校准";
    }
    if ("issued".equals(status)) {
      return who + "绩效结果待确认";
    }
    return who + "待办";
  }
}

File: server/modules/home/home.service.ts
import { Injectable, Inject } from '@nestjs/common';
import { DRIZZLE_DATABASE } from '@lark-apaas/fullstack-nestjs-core';
import type { AppDatabase } from '@server/database/app-database';
import { eq, or, and, count, inArray, desc, gte, lt, type SQL } from 'drizzle-orm';
import { performanceRecord } from '../../database/schema';
import { MenuPermissionService } from '../menu-permission/menu-permission.service';
import type { TodoItem, PerformanceOverview, HomeActionCounts } from '@shared/api.interface';

const ACTIVE_STATUSES = [
  'template_selection',
  'goal_setting',
  'goal_pending_review',
  'goal_rejected',
  'self_review',
  'manager_review',
  'dual_manager_review',
  'dotted_manager_review',
  'final_review',
] as const;

function resolveYearMonth(year?: number, month?: number): { year: number; month: number } {
  const now = new Date();
  const y = year != null && Number.isFinite(year) ? Math.floor(year) : now.getFullYear();
  const m = month != null && Number.isFinite(month) ? Math.min(12, Math.max(1, Math.floor(month))) : now.getMonth() + 1;
  return { year: y, month: m };
}

function monthUpdatedRange(year: number, month: number): { start: Date; endExclusive: Date } {
  const start = new Date(year, month - 1, 1, 0, 0, 0, 0);
  const endExclusive = new Date(year, month, 1, 0, 0, 0, 0);
  return { start, endExclusive };
}

@Injectable()
export class HomeService {
  constructor(
    @Inject(DRIZZLE_DATABASE) private readonly db: AppDatabase,
    private readonly menuPermissionService: MenuPermissionService,
  ) {}

  private async getUserRole(userId: string): Promise<string> {
    return this.menuPermissionService.getUserRole(userId);
  }

  private isAdminRole(role: string): boolean {
    return role === 'super_admin' || role === 'admin';
  }

  /** 当前用户可见的绩效范围条件（管理员看全量） */
  private async userScopeCondition(userId: string): Promise<SQL | undefined> {
    const role = await this.getUserRole(userId);
    if (this.isAdminRole(role)) {
      return undefined;
    }
    return or(
      eq(performanceRecord.employeeId, userId),
      eq(performanceRecord.managerId, userId),
      eq(performanceRecord.dottedManagerId, userId),
    );
  }

  private monthCondition(year: number, month: number): SQL {
    const { start, endExclusive } = monthUpdatedRange(year, month);
    return and(
      gte(performanceRecord.updatedAt, start),
      lt(performanceRecord.updatedAt, endExclusive),
    )!;
  }

  async getTodos(
    userId: string,
    opts?: { year?: number; month?: number },
  ): Promise<{ items: TodoItem[] }> {
    const { year, month } = resolveYearMonth(opts?.year, opts?.month);
    const monthCond = this.monthCondition(year, month);
    const role = await this.getUserRole(userId);
    const admin = this.isAdminRole(role);

    const clauses: SQL[] = [
      and(
        eq(performanceRecord.status, 'template_selection'),
        eq(performanceRecord.employeeId, userId),
      ),
      and(eq(performanceRecord.status, 'goal_setting'), eq(performanceRecord.employeeId, userId)),
      and(eq(performanceRecord.status, 'goal_rejected'), eq(performanceRecord.employeeId, userId)),
      and(
        eq(performanceRecord.status, 'goal_pending_review'),
        or(
          eq(performanceRecord.managerId, userId),
          eq(performanceRecord.dottedManagerId, userId),
        ),
      ),
      and(eq(performanceRecord.status, 'self_review'), eq(performanceRecord.employeeId, userId)),
      and(eq(performanceRecord.status, 'manager_review'), eq(performanceRecord.managerId, userId)),
      and(
        eq(performanceRecord.status, 'dual_manager_review'),
        or(
          eq(performanceRecord.managerId, userId),
          eq(performanceRecord.dottedManagerId, userId),
        ),
      ),
      and(
        eq(performanceRecord.status, 'dotted_manager_review'),
        eq(performanceRecord.dottedManagerId, userId),
      ),
    ];
    if (admin) {
      clauses.push(eq(performanceRecord.status, 'final_review'));
    }

    const rows = await this.db
      .select({
        id: performanceRecord.id,
        period: performanceRecord.period,
        status: performanceRecord.status,
      })
      .from(performanceRecord)
      .where(and(or(...clauses), monthCond))
      .orderBy(desc(performanceRecord.updatedAt));

    const items: TodoItem[] = rows.map((r) => {
      const period = r.period;
      switch (r.status) {
        case 'template_selection':
          return { id: r.id, period, type: 'template_selection', title: `${period} 待选择绩效模板` };
        case 'goal_setting':
          return { id: r.id, period, type: 'goal_setting', title: `${period} 目标设定` };
        case 'goal_rejected':
          return { id: r.id, period, type: 'goal_rejected', title: `${period} 目标被驳回，请修改` };
        case 'goal_pending_review':
          return { id: r.id, period, type: 'goal_pending_review', title: `${period} 待审核目标` };
        case 'self_review':
          return { id: r.id, period, type: 'self_review', title: `${period} 绩效自评` };
        case 'manager_review':
          return { id: r.id, period, type: 'manager_review', title: `${period} 直属上级评分` };
        case 'dual_manager_review':
          return { id: r.id, period, type: 'dual_manager_review', title: `${period} 上级与虚线上级评分` };
        case 'dotted_manager_review':
          return { id: r.id, period, type: 'dotted_manager_review', title: `${period} 虚线上级评分` };
        case 'final_review':
          return { id: r.id, period, type: 'final_review', title: `${period} 待终审/校准` };
        default:
          return { id: r.id, period, type: 'self_review', title: `${period} 待办` };
      }
    });

    return { items };
  }

  async getOverview(
    userId: string,
    opts?: { year?: number; month?: number },
  ): Promise<PerformanceOverview> {
    const { year, month } = resolveYearMonth(opts?.year, opts?.month);
    const monthCond = this.monthCondition(year, month);
    const scope = await this.userScopeCondition(userId);

    const baseWhere = scope ? and(scope, monthCond) : monthCond;

    const totalResult = await this.db
      .select({ count: count() })
      .from(performanceRecord)
      .where(baseWhere);

    const pendingWhere = and(
      baseWhere,
      inArray(performanceRecord.status, [...ACTIVE_STATUSES] as unknown as string[]),
    );

    const pendingResult = await this.db
      .select({ count: count() })
      .from(performanceRecord)
      .where(pendingWhere);

    const completedWhere = and(baseWhere, eq(performanceRecord.status, 'completed'));

    const completedResult = await this.db
      .select({ count: count() })
      .from(performanceRecord)
      .where(completedWhere);

    const rejectedWhere = and(baseWhere, eq(performanceRecord.status, 'goal_rejected'));

    const rejectedResult = await this.db
      .select({ count: count() })
      .from(performanceRecord)
      .where(rejectedWhere);

    const total = Number((totalResult[0] as { count: unknown }).count);
    const pending = Number((pendingResult[0] as { count: unknown }).count);
    const completed = Number((completedResult[0] as { count: unknown }).count);
    const rejected = Number((rejectedResult[0] as { count: unknown }).count);

    return { total, pending, completed, rejected, year, month };
  }

  async getActionCounts(
    userId: string,
    opts?: { year?: number; month?: number },
  ): Promise<HomeActionCounts> {
    const { year, month } = resolveYearMonth(opts?.year, opts?.month);
    const monthCond = this.monthCondition(year, month);
    const role = await this.getUserRole(userId);
    const admin = this.isAdminRole(role);

    const mgrScore = await this.db
      .select({ count: count() })
      .from(performanceRecord)
      .where(
        and(
          eq(performanceRecord.status, 'manager_review'),
          eq(performanceRecord.managerId, userId),
          monthCond,
        ),
      );

    const dualScore = await this.db
      .select({ count: count() })
      .from(performanceRecord)
      .where(
        and(
          eq(performanceRecord.status, 'dual_manager_review'),
          or(
            eq(performanceRecord.managerId, userId),
            eq(performanceRecord.dottedManagerId, userId),
          ),
          monthCond,
        ),
      );

    const dottedScore = await this.db
      .select({ count: count() })
      .from(performanceRecord)
      .where(
        and(
          eq(performanceRecord.status, 'dotted_manager_review'),
          eq(performanceRecord.dottedManagerId, userId),
          monthCond,
        ),
      );

    const approveGoal = await this.db
      .select({ count: count() })
      .from(performanceRecord)
      .where(
        and(
          eq(performanceRecord.status, 'goal_pending_review'),
          or(
            eq(performanceRecord.managerId, userId),
            eq(performanceRecord.dottedManagerId, userId),
          ),
          monthCond,
        ),
      );

    const needScore =
      Number((mgrScore[0] as { count: unknown }).count) +
      Number((dualScore[0] as { count: unknown }).count) +
      Number((dottedScore[0] as { count: unknown }).count);
    const needApproveGoal = Number((approveGoal[0] as { count: unknown }).count);

    let needFinalReview = 0;
    if (admin) {
      const fr = await this.db
        .select({ count: count() })
        .from(performanceRecord)
        .where(and(eq(performanceRecord.status, 'final_review'), monthCond));
      needFinalReview = Number((fr[0] as { count: unknown }).count);
    }

    return { needScore, needApproveGoal, needFinalReview };
  }
}
